unistore crash on data import

## Bug Report

```
panic: runtime error: index out of range [-1]

goroutine 385 [running]:
github.com/pingcap/tidb/pkg/store/mockstore/unistore/lockstore.(*arena).get(0xc0c4262460?, 0xda8d7a?, 0x3a0b2ff?)
	/workspace/source/tidb/pkg/store/mockstore/unistore/lockstore/arena.go:79 +0x210
github.com/pingcap/tidb/pkg/store/mockstore/unistore/lockstore.(*MemStore).newNode(0xc001210060, 0xc001192000, {0xc0c427a06c, 0x13, 0xc001106e58?}, {0xc1122b6000, 0xda8d4f, 0xc001d41340?}, 0x1)
	/workspace/source/tidb/pkg/store/mockstore/unistore/lockstore/lockstore.go:366 +0x38e
github.com/pingcap/tidb/pkg/store/mockstore/unistore/lockstore.(*MemStore).PutWithHint(0xc001210060, {0xc0c427a06c, 0x13, 0x13}, {0xc1122b6000, 0xda8d4f, 0xda8d4f}, 0xc001106e58?)
	/workspace/source/tidb/pkg/store/mockstore/unistore/lockstore/lockstore.go:316 +0x245
github.com/pingcap/tidb/pkg/store/mockstore/unistore/tikv.writeLockWorker.run({0xc0011a6540?, 0xc00104d480?})
	/workspace/source/tidb/pkg/store/mockstore/unistore/tikv/write.go:170 +0x431
created by github.com/pingcap/tidb/pkg/store/mockstore/unistore/tikv.(*dbWriter).Open in goroutine 1
	/workspace/source/tidb/pkg/store/mockstore/unistore/tikv/write.go:207 +0x147
```

### 1. Minimal reproduce step (Required)

Load sql file created with mysqldump with `mysql -e source some_file.sql`

Config:
```yaml
port = 3306
path = "/var/tidb"
txn-entry-size-limit = 16777216
allow-expression-index = true
```

### 2. What did you expect to see? (Required)

### 3. What did you see instead (Required)

### 4. What is your TiDB version? (Required)

v8.5.1 with unistore

This doesn't need a container:
```
./bin/tidb-server -config <(echo -en "[performance]\ntxn-entry-size-limit = 125829120\n")
```

```
$ mysql -h 127.0.0.1 -P 4000 -u root test
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 2097156
Server version: 8.0.11-TiDB-v8.5.1 TiDB Server (Apache License 2.0) Community Edition, MySQL 8.0 compatible

Copyright (c) 2000, 2025, Oracle and/or its affiliates.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql-8.0.11-TiDB-v8.5.1> INSERT INTO t1(b) VALUES(REPEAT('x',10000000));
ERROR 2013 (HY000): Lost connection to MySQL server during query
No connection. Trying to reconnect...
ERROR 2003 (HY000): Can't connect to MySQL server on '127.0.0.1:4000' (111)
ERROR: 
Can't connect to the server

mysql-not_connected> 
```

```
[2025/02/21 15:04:18.732 +01:00] [INFO] [2pc.go:693] ["[BIG_TXN]"] [session=2097156] ["key sample"=74800000000000006e5f728000000000007531] [size=10000033] [keys=1] [puts=1] [dels=0] [locks=0] [checks=0] [txnStartTS=456169005703233536]
panic: runtime error: index out of range [-1]

goroutine 44 [running]:
github.com/pingcap/tidb/pkg/store/mockstore/unistore/lockstore.(*arena).get(0xc01b16e9b0?, 0x9896f4?, 0x202aca5?)
	/home/dvaneeden/dev/pingcap/tidb/pkg/store/mockstore/unistore/lockstore/arena.go:79 +0x210
github.com/pingcap/tidb/pkg/store/mockstore/unistore/lockstore.(*MemStore).newNode(0xc0017b8060, 0xc001bae000, {0xc01b16a06c, 0x13, 0x13?}, {0xc01df9e000, 0x9896c9, 0xc001e988c0?}, 0x1)
	/home/dvaneeden/dev/pingcap/tidb/pkg/store/mockstore/unistore/lockstore/lockstore.go:366 +0x38e
github.com/pingcap/tidb/pkg/store/mockstore/unistore/lockstore.(*MemStore).PutWithHint(0xc0017b8060, {0xc01b16a06c, 0x13, 0x13}, {0xc01df9e000, 0x9896c9, 0x9896c9}, 0x3?)
	/home/dvaneeden/dev/pingcap/tidb/pkg/store/mockstore/unistore/lockstore/lockstore.go:316 +0x245
github.com/pingcap/tidb/pkg/store/mockstore/unistore/tikv.writeLockWorker.run({0xc000f9c380?, 0xc0008673c0?})
	/home/dvaneeden/dev/pingcap/tidb/pkg/store/mockstore/unistore/tikv/write.go:170 +0x431
created by github.com/pingcap/tidb/pkg/store/mockstore/unistore/tikv.(*dbWriter).Open in goroutine 1
	/home/dvaneeden/dev/pingcap/tidb/pkg/store/mockstore/unistore/tikv/write.go:207 +0x147
```

With some extra debug logging added:

```patch
diff --git a/pkg/store/mockstore/unistore/lockstore/arena.go b/pkg/store/mockstore/unistore/lockstore/arena.go
index bf02e5020e..649316047e 100644
--- a/pkg/store/mockstore/unistore/lockstore/arena.go
+++ b/pkg/store/mockstore/unistore/lockstore/arena.go
@@ -76,6 +76,7 @@ func (a *arena) get(addr arenaAddr, size int) []byte {
        if addr.blockIdx() >= len(a.blocks) {
                log.S().Fatalf("arena.get out of range. len(blocks)=%v, addr.blockIdx()=%v, addr.blockOffset()=%v, size=%v", len(a.blocks), addr.blockIdx(), addr.blockOffset(), size)
        }
+       log.S().Infof("arena.get: len(blocks)=%v, addr.blockIdx()=%v, addr.blockOffset()=%v, size=%v", len(a.blocks), addr.blockIdx(), addr.blockOffset(), size)
        return a.blocks[addr.blockIdx()].get(addr.blockOffset(), size)
 }
 
```

```
[2025/02/21 15:22:15.832 +01:00] [INFO] [arena.go:79] ["arena.get: len(blocks)=1, addr.blockIdx()=0, addr.blockOffset()=144, size=16"]
[2025/02/21 15:22:15.832 +01:00] [INFO] [arena.go:79] ["arena.get: len(blocks)=1, addr.blockIdx()=0, addr.blockOffset()=144, size=59"]
[2025/02/21 15:22:15.832 +01:00] [INFO] [arena.go:79] ["arena.get: len(blocks)=1, addr.blockIdx()=0, addr.blockOffset()=144, size=16"]
[2025/02/21 15:22:15.832 +01:00] [INFO] [arena.go:79] ["arena.get: len(blocks)=1, addr.blockIdx()=0, addr.blockOffset()=144, size=59"]
[2025/02/21 15:22:15.844 +01:00] [INFO] [2pc.go:693] ["[BIG_TXN]"] [session=2097154] ["key sample"=74800000000000006e5f728000000000015f91] [size=10000033] [keys=1] [puts=1] [dels=0] [locks=0] [checks=0] [txnStartTS=456169288061681664]
[2025/02/21 15:22:15.849 +01:00] [INFO] [arena.go:79] ["arena.get: len(blocks)=2, addr.blockIdx()=-1, addr.blockOffset()=0, size=10000132"]
panic: runtime error: index out of range [-1]

goroutine 83 [running]:
github.com/pingcap/tidb/pkg/store/mockstore/unistore/lockstore.(*arena).get(0xc01339e0f0, 0x0, 0x989704)
	/home/dvaneeden/dev/pingcap/tidb/pkg/store/mockstore/unistore/lockstore/arena.go:80 +0x393
github.com/pingcap/tidb/pkg/store/mockstore/unistore/lockstore.(*MemStore).newNode(0xc00121a150, 0xc0010a6000, {0xc0133ab06c, 0x13, 0x13?}, {0xc016c1e000, 0x9896c9, 0xc00119fc00?}, 0x3)
	/home/dvaneeden/dev/pingcap/tidb/pkg/store/mockstore/unistore/lockstore/lockstore.go:366 +0x38e
github.com/pingcap/tidb/pkg/store/mockstore/unistore/lockstore.(*MemStore).PutWithHint(0xc00121a150, {0xc0133ab06c, 0x13, 0x13}, {0xc016c1e000, 0x9896c9, 0x9896c9}, 0x0?)
	/home/dvaneeden/dev/pingcap/tidb/pkg/store/mockstore/unistore/lockstore/lockstore.go:316 +0x245
github.com/pingcap/tidb/pkg/store/mockstore/unistore/tikv.writeLockWorker.run({0xc001242460?, 0xc0011f8040?})
	/home/dvaneeden/dev/pingcap/tidb/pkg/store/mockstore/unistore/tikv/write.go:170 +0x431
created by github.com/pingcap/tidb/pkg/store/mockstore/unistore/tikv.(*dbWriter).Open in goroutine 1
	/home/dvaneeden/dev/pingcap/tidb/pkg/store/mockstore/unistore/tikv/write.go:207 +0x147
```

--- a/pkg/executor/insert_test.go
+++ b/pkg/executor/insert_test.go
@@ -703,3 +703,14 @@ func TestInsertNullInNonStrictMode(t *testing.T) {
 	tk.MustExec("insert ignore t1 VALUES (4, 4) ON DUPLICATE KEY UPDATE col1 = null")
 	tk.MustQuery("select * from t1").Check(testkit.RowsWithSep("|", "1|", "2|", "3|", "4|", "5|"))
 }
+
+func TestInsertLargeRow(t *testing.T) {
+	store := testkit.CreateMockStore(t)
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+	tk.MustExec("create table t (id int primary key, b longtext)")
+	tk.MustExec("set tidb_txn_entry_size_limit = 1<<23")
+	// the unistore arena blocksize is 8MB (8388608 bytes), so Unistore cannot handle larger rows than that!
+	// since a row cannot span multiple arena blocks.
+	tk.MustExec("insert into t values (1, REPEAT('t',8388493))")
+}

--- a/pkg/executor/insert_test.go
+++ b/pkg/executor/insert_test.go
@@ -712,5 +712,5 @@ func TestInsertLargeRow(t *testing.T) {
 	tk.MustExec("set tidb_txn_entry_size_limit = 1<<23")
 	// the unistore arena blocksize is 8MB (8388608 bytes), so Unistore cannot handle larger rows than that!
 	// since a row cannot span multiple arena blocks.
-	tk.MustExec("insert into t values (1, REPEAT('t',8388493))")
+	tk.MustContainErrMsg("insert into t values (1, REPEAT('t',8388493))", "Lock entry too big")
 }

--- a/pkg/executor/insert_test.go
+++ b/pkg/executor/insert_test.go
@@ -17,6 +17,7 @@ package executor_test
 import (
 	"fmt"
 	"strconv"
+	"strings"
 	"testing"
 	"time"
 
@@ -707,10 +708,14 @@ func TestInsertNullInNonStrictMode(t *testing.T) {
 func TestInsertLargeRow(t *testing.T) {
 	store := testkit.CreateMockStore(t)
 	tk := testkit.NewTestKit(t, store)
+	ver := tk.MustQuery("select tidb_version()").Rows()[0][0].(string)
+	if !strings.Contains(ver, "Store: unistore") {
+		t.Skipf("Only support 'Store: unistore'\n%s", ver)
+	}
 	tk.MustExec("use test")
 	tk.MustExec("create table t (id int primary key, b longtext)")
 	tk.MustExec("set tidb_txn_entry_size_limit = 1<<23")
 	// the unistore arena blocksize is 8MB (8388608 bytes), so Unistore cannot handle larger rows than that!
 	// since a row cannot span multiple arena blocks.
-	tk.MustContainErrMsg("insert into t values (1, REPEAT('t',8388493))", "Lock entry too big")
+	tk.MustContainErrMsg("insert into t values (1, REPEAT('t',8388493))", "unistore lock entry too big")
 }

--- a/pkg/store/mockstore/unistore/lockstore/lockstore.go
+++ b/pkg/store/mockstore/unistore/lockstore/lockstore.go
@@ -353,6 +353,10 @@ func (ls *MemStore) replace(key, v []byte, hint *Hint, old *node) {
 	ls.getArena().free(old.addr)
 }
 
+func (ls *MemStore) MaxEntrySize() int {
+	return ls.getArena().blockSize - nodeHeaderSize - ls.getHeight()*8
+}
+
 func (ls *MemStore) newNode(arena *arena, key []byte, v []byte, height int) *node {
 	// The base level is already allocated in the node struct.
 	nodeSize := nodeHeaderSize + height*8 + len(key) + len(v)

--- a/pkg/store/mockstore/unistore/lockstore/lockstore.go
+++ b/pkg/store/mockstore/unistore/lockstore/lockstore.go
@@ -353,6 +353,8 @@ func (ls *MemStore) replace(key, v []byte, hint *Hint, old *node) {
 	ls.getArena().free(old.addr)
 }
 
+// MaxEntrySize will return the maximum entry size for the MemStore
+// Any entry larger than this will likely result in failure.
 func (ls *MemStore) MaxEntrySize() int {
 	return ls.getArena().blockSize - nodeHeaderSize - ls.getHeight()*8
 }

--- a/pkg/store/mockstore/unistore/tikv/write.go
+++ b/pkg/store/mockstore/unistore/tikv/write.go
@@ -16,6 +16,8 @@ package tikv
 
 import (
 	"bytes"
+	"errors"
+	"fmt"
 	"math"
 	"sync"
 	"sync/atomic"
@@ -166,6 +168,11 @@ func (w writeLockWorker) run() {
 					// Ignore if the key doesn't exist
 					ls.DeleteWithHint(entry.Key.UserKey, hint)
 				default:
+					// Make sure it fits into an arena block.
+					if len(entry.Key.UserKey)+len(entry.Value) > ls.MaxEntrySize() {
+						batch.err = errors.New(fmt.Sprintf("Lock entry too big %d > %d", len(entry.Key.UserKey)+len(entry.Value), ls.MaxEntrySize()))
+						break
+					}
 					insertCnt++
 					ls.PutWithHint(entry.Key.UserKey, entry.Value, hint)
 				}

--- a/pkg/store/mockstore/unistore/tikv/write.go
+++ b/pkg/store/mockstore/unistore/tikv/write.go
@@ -16,7 +16,6 @@ package tikv
 
 import (
 	"bytes"
-	"errors"
 	"fmt"
 	"math"
 	"sync"
@@ -170,7 +169,7 @@ func (w writeLockWorker) run() {
 				default:
 					// Make sure it fits into an arena block.
 					if len(entry.Key.UserKey)+len(entry.Value) > ls.MaxEntrySize() {
-						batch.err = errors.New(fmt.Sprintf("Lock entry too big %d > %d", len(entry.Key.UserKey)+len(entry.Value), ls.MaxEntrySize()))
+						batch.err = fmt.Errorf("Lock entry too big %d > %d", len(entry.Key.UserKey)+len(entry.Value), ls.MaxEntrySize())
 						break
 					}
 					insertCnt++

--- a/pkg/store/mockstore/unistore/tikv/write.go
+++ b/pkg/store/mockstore/unistore/tikv/write.go
@@ -169,7 +169,7 @@ func (w writeLockWorker) run() {
 				default:
 					// Make sure it fits into an arena block.
 					if len(entry.Key.UserKey)+len(entry.Value) > ls.MaxEntrySize() {
-						batch.err = fmt.Errorf("Unistore lock entry too big %d > %d", len(entry.Key.UserKey)+len(entry.Value), ls.MaxEntrySize())
+						batch.err = fmt.Errorf("unistore lock entry too big %d > %d", len(entry.Key.UserKey)+len(entry.Value), ls.MaxEntrySize())
 						break
 					}
 					insertCnt++

--- a/pkg/store/mockstore/unistore/tikv/write.go
+++ b/pkg/store/mockstore/unistore/tikv/write.go
@@ -169,7 +169,7 @@ func (w writeLockWorker) run() {
 				default:
 					// Make sure it fits into an arena block.
 					if len(entry.Key.UserKey)+len(entry.Value) > ls.MaxEntrySize() {
-						batch.err = fmt.Errorf("Lock entry too big %d > %d", len(entry.Key.UserKey)+len(entry.Value), ls.MaxEntrySize())
+						batch.err = fmt.Errorf("Unistore lock entry too big %d > %d", len(entry.Key.UserKey)+len(entry.Value), ls.MaxEntrySize())
 						break
 					}
 					insertCnt++

